### PR TITLE
[Feature] Add test-time ensemble for images and videos and support ensemble in BasicVSR series

### PR DESCRIPTION
**Pull request by ckkelvinchan**
## Motivation
1. Support spatial and temporal ensemble (e.g., rotation, reflection), which help improve the performance and are used by many methods.
2. Support ensemble for BasicVSR series.

## Modification
1. Add `mmedit/models/common/ensemble.py` for the ensemble
2. Modify `mmedit/models/restorers/basicvsr.py` to support ensemble for BasicVSR, IconVSR, and BasicVSR++
3. Modify `configs/restorers/basicvsr_plusplus/basicvsr_plusplus_c128n25_*.py`

### CODE REVIEW

File: mmedit/models/common/ensemble.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        is_single_image = False
+        if imgs.ndim == 4:
+            if self.is_temporal_ensemble:
+                raise ValueError('"is_temporal_ensemble" must be False if '
+                                 'the input is an image.')
+            is_single_image = True
+            imgs = imgs.unsqueeze(1)
+
+        if mode == 'vertical':
+            imgs = imgs.flip(4).clone()
```
Comment (Yshuo-Li) [inline mmedit/models/common/ensemble.py:35]: Only support video
or
Support video and single image

how about 
`flip(-1)`
`transpose(-2, -1)`

Comment (Yshuo-Li) [inline mmedit/models/common/ensemble.py:35]: An issue

**Review by Yshuo-Li — COMMENTED** [review]
Only support video
or
Support video and single image

how about
`flip(-1)`
`transpose(-2, -1)`

File: mmedit/models/common/ensemble.py (hunk 1 above)
Comment (ckkelvinchan) [reply mmedit/models/common/ensemble.py:35]: If the input is an image (n, c, h, w), it is unsqueezed to (n, 1, c, h, w). So it should be fine.

**Review by Yshuo-Li — APPROVED** [review]

**Comment by Yshuo-Li** [comment]
Need resolve conflicts

File: mmedit/models/common/ensemble.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        Tensor: Model outputs given the ensembles.
+    """
+
+    def __init__(self, is_temporal_ensemble=False):
+
+        super().__init__()
+
+        self.is_temporal_ensemble = is_temporal_ensemble
+
+    def _transform(self, imgs, mode):
```
Comment (Yshuo-Li) [inline mmedit/models/common/ensemble.py:25]: Need docstring

File: mmedit/models/common/ensemble.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            imgs = imgs.flip(3).clone()
+        elif mode == 'transpose':
+            imgs = imgs.permute(0, 1, 2, 4, 3).clone()
+
+        if is_single_image:
+            imgs = imgs.squeeze(1)
+
+        return imgs
+
+    def spatial_ensemble(self, imgs, model):
```
Comment (Yshuo-Li) [inline mmedit/models/common/ensemble.py:46]: Need docstring

File: mmedit/models/common/ensemble.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+        if mode == 'vertical':
+            imgs = imgs.flip(4).clone()
+        elif mode == 'horizontal':
+            imgs = imgs.flip(3).clone()
+        elif mode == 'transpose':
+            imgs = imgs.permute(0, 1, 2, 4, 3).clone()
+
+        if is_single_image:
+            imgs = imgs.squeeze(1)
```
Comment (Yshuo-Li) [inline mmedit/models/common/ensemble.py:42]: I think squeeze() should be placed before flip() and permute()

File: mmedit/models/common/ensemble.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                output_list[i] = self._transform(output_list[i], 'horizontal')
+            if (i % 4) % 2 == 1:
+                output_list[i] = self._transform(output_list[i], 'vertical')
+
+        outputs = torch.stack(output_list, dim=0)
+        outputs = outputs.mean(dim=0, keepdim=False)
+
+        return outputs.to(imgs.device)
+
+    def forward(self, imgs, model):
```
Comment (Yshuo-Li) [inline mmedit/models/common/ensemble.py:65]: Need docstring

File: configs/restorers/basicvsr_plusplus/basicvsr_plusplus_c128n25_600k_ntire_decompress_track1.py [hunk 6]
```diff
@@ -11,7 +11,9 @@
         spynet_pretrained='https://download.openmmlab.com/mmediting/restorers/'
         'basicvsr/spynet_20210409-c6c1bd09.pth',
         cpu_cache_length=100),
-    pixel_loss=dict(type='CharbonnierLoss', loss_weight=1.0, reduction='mean'))
+    pixel_loss=dict(type='CharbonnierLoss', loss_weight=1.0, reduction='mean'),
+    ensemble=dict(is_temporal_ensemble=False),
```
Comment (Yshuo-Li) [inline configs/restorers/basicvsr_plusplus/basicvsr_plusplus_c128n25_600k_ntire_decompress_track1.py:15]: How about working as a register
` ensemble=dict(type='SpatialTemporalEnsemble', is_temporal_ensemble=False)`

File: mmedit/models/common/ensemble.py (hunk 4 above)
Comment (ckkelvinchan) [reply mmedit/models/common/ensemble.py:42]: `unsqueeze` is to make sure the size of the tensor is of `(n, t, c, h, w)`. Therefore the operations do not need to distinguish `image` and `video` inputs.

After that it is squeezed.

**Review by Yshuo-Li — APPROVED** [review]

**Pull request merged**